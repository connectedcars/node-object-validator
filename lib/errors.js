function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
