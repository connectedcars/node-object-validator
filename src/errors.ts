export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
