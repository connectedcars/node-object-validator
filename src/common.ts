import { RequiredFail, ValidationFailure, ValidationsError } from './errors'

// https://stackoverflow.com/questions/51651499/typescript-what-is-a-naked-type-parameter
// https://2ality.com/2019/07/testing-static-types.html
// Wrapping the types in an tuple force a specific type instead of allow any in the union
export type AssertEqual<T, Expected> = [T, Expected] extends [Expected, T] ? true : false

export type ReturnEqual<T, C> = [T, C] extends [C, T] ? C : never

// https://github.com/microsoft/TypeScript/issues/41746
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BaseObject = Record<string, any>

// TODO: Give better name
export function isValidType<T>(value: unknown, errors: ValidationFailure[]): value is T {
  return errors.length === 0
}

export type CodeGenResult = [{ [key: string]: unknown }, string[], string[]]

export type ValidatorBaseOptions = {
  optimize?: boolean
  required?: boolean
  nullable?: boolean
  earlyFail?: boolean
}

/**
 * @typedef ValidatorOptions
 * @property {boolean} [earlyFail=false] Stop validation on first failure (default: false)
 * @property {boolean} [optimize=true] Generate an optimized function for doing the validation (default: true)
 */
export interface ValidatorOptions {
  earlyFail?: boolean
  optimize?: boolean
}

export interface ValidateOptions {
  earlyFail?: boolean
  optimized?: boolean
}

export interface ValidatorExportOptions {
  language?: string
  types?: boolean
}

export function isValidator(value: unknown): value is ValidatorBase {
  if (value instanceof ValidatorBase) {
    return true
  }
  return false
}

export function generateOptionsString(options: ValidatorBaseOptions, defaults: Required<ValidatorBaseOptions>): string {
  const selectedOptions: string[] = []
  if (options.required !== undefined && options.required !== defaults.required) {
    selectedOptions.push(`required: ${options.required}`)
  }
  if (options.nullable !== undefined && options.nullable !== defaults.nullable) {
    selectedOptions.push(`nullable: ${options.nullable}`)
  }
  if (options.earlyFail !== undefined && options.earlyFail !== defaults.earlyFail) {
    selectedOptions.push(`earlyFail: ${options.earlyFail}`)
  }
  if (options.optimize !== undefined && options.optimize !== defaults.optimize) {
    selectedOptions.push(`optimize: ${options.optimize}`)
  }
  return selectedOptions.length > 0 ? `{ ${selectedOptions.join(', ')} }` : ''
}

export abstract class ValidatorBase<T = unknown> {
  public tsType!: T
  public required: boolean
  public nullable: boolean
  public earlyFail: boolean

  protected codeGenId = 1
  protected optimizedValidate: ((value: unknown, context?: string) => ValidationFailure[]) | null
  protected optionsString: string

  public constructor(options?: ValidatorBaseOptions) {
    const defaults = { required: true, nullable: false, earlyFail: false, optimize: true }
    const mergedOptions = { ...defaults, ...options }

    this.optionsString = options ? generateOptionsString(options, defaults) : ''

    this.required = mergedOptions.required
    this.earlyFail = mergedOptions.earlyFail
    this.nullable = mergedOptions.nullable
    this.optimizedValidate = null
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public AssertType<C extends T, X extends AssertEqual<T, C>>(): void {
    // Only used for type assertions
  }

  public isValid<C extends T>(obj: unknown): obj is ReturnEqual<T, C> {
    const errors = this.validate(obj)
    return errors.length === 0
  }

  public isType<C extends T>(obj: unknown, errors: ValidationFailure[]): obj is ReturnEqual<T, C> {
    return errors.length === 0
  }

  public cast<C extends T>(obj: unknown): ReturnEqual<T, C> {
    const errors = this.validate(obj)
    if (this.isType<C>(obj, errors)) {
      return obj
    } else {
      throw new ValidationsError('One of more validations failed', errors)
    }
  }

  public validate(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    if (options?.optimized !== false && this.optimizedValidate !== null) {
      return this.optimizedValidate(value, context)
    }
    if (this.nullable && value === null) {
      return []
    }
    if (value === undefined) {
      return this.required ? [new RequiredFail(`Is required`, value, context)] : []
    }
    return this.validateValue(value, context, { earlyFail: this.earlyFail, optimized: false, ...options })
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, { key: (context ? \`\${context}['${context}']\` : \`${context}\`) }` : ', context'
    const validatorName = `validator${id()}`
    const declarations = [`const ${validatorName} = ${validatorRef}`]
    // prettier-ignore
    const code: string[] = [
      ...this.nullCheckWrap([
        `  errors.push(...${validatorName}.validateValue(${valueRef}${contextStr}))`,
      ],valueRef, contextStr),
      ...(this.earlyFail || earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  protected nullCheckWrap(code: string[], valueRef: string, contextStr: string, skip = false): string[] {
    if (skip) {
      return code
    }

    const result: string[] = []
    if (this.required) {
      if (this.nullable) {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(`  if (${valueRef} !== null) {`)
        result.push(...code)
        result.push(`  }`)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      }
    } else {
      if (this.nullable) {
        result.push(`if (${valueRef} != null) {`)
        result.push(...code)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`}`)
      }
    }
    return result
  }

  protected optimize(schema?: unknown): void {
    const [imports, declarations, code] = this.codeGen(
      'value',
      'schema',
      () => {
        return this.codeGenId++
      },
      undefined,
      this.earlyFail
    )
    const functionBody = [
      ...Object.keys(imports).map(i => `const ${i} = imports['${i}']`),
      ...declarations,
      `return (value, context) => {`,
      `  const generatedFunction = true`,
      `  const errors = []`,
      ...code.map(l => `  ${l}`),
      `  return errors`,
      `}`
    ].join('\n')

    try {
      const functionGenerator = new Function('imports', 'schema', functionBody)
      const validateFunction = functionGenerator(imports, {
        schema,
        validateValue: this.validateValue.bind(this)
      })
      this.optimizedValidate = validateFunction
    } catch (e) {
      throw new Error(`Failed to compile optimized function(${e.message}):\n${functionBody}`)
    }
  }

  public abstract toString(options?: ValidatorExportOptions): string

  protected abstract validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[]
}
