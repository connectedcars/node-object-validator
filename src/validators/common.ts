import { ValidationErrorContext } from '../errors'

export type Schema = {
  [key: string]: Validator
}

export interface Validator {
  schema?: Schema
  validate(value: unknown, context?: ValidationErrorContext): Error | null
}
