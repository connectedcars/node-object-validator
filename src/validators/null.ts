import {
  CodeGenResult,
  generateOptionsString,
  ValidatorBase,
  ValidatorBaseOptions,
  ValidatorExportOptions,
  ValidatorOptions
} from '../common'
import { NotNullFail, RequiredFail, ValidationFailure } from '../errors'

export function isNull(value: unknown, context?: string): value is null {
  const errors = validateNull(value, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateNull(value: unknown, context?: string): ValidationFailure[] {
  if (value !== null) {
    return [new NotNullFail(`Must be an null`, value, context)]
  }
  return []
}

export abstract class NullValidator<O = never> extends ValidatorBase<null | O> {
  public constructor(options?: ValidatorBaseOptions) {
    super({ ...options })
    this.optionsString = options
      ? generateOptionsString(options, {
          required: true,
          nullable: false,
          earlyFail: false,
          optimize: true
        })
      : ''
    if (options?.optimize !== false) {
      this.optimize()
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      `if (${localValueRef} !== undefined) {`,
      `  if (${localValueRef} !== null) {`,
      `    errors.push(new NotNullFail(\`Must be an null\`, ${localValueRef}${contextStr}))`,
      `  }`,
      ...(this.required ? [
      `} else {`,
      `  errors.push(new RequiredFail(\`Is required\`, ${localValueRef}${contextStr}))`] : []),
      '}',
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),

    ]
    return [
      {
        NotNullFail: NotNullFail,
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string): ValidationFailure[] {
    return validateNull(value, context)
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = `null`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        throw new Error('Rust not supported yet')
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredNull extends NullValidator {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: true })
  }
}

export class OptionalNull extends NullValidator<undefined> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false })
  }
}
