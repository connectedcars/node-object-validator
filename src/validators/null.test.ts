import { AssertEqual } from '../common'
import { NotNullFail, RequiredFail } from '../errors'
import { isNull, OptionalNull, RequiredNull, validateNull } from './null'

describe('Null', () => {
  describe('validateNull', () => {
    it('requires value to be null', () => {
      const value = null as unknown
      expect(validateNull(value)).toStrictEqual([])
    })
  })

  describe('isNull', () => {
    it('should cast value to null', () => {
      const value = null as unknown
      if (isNull(value)) {
        expect(true as AssertEqual<typeof value, null>).toEqual(true)
      } else {
        fail('did not validate but should')
      }
    })

    it('should fail validation', () => {
      const value = 'string' as unknown
      expect(isNull(value)).toEqual(false)
    })
  })

  describe('RequiredNull', () => {
    it('should return a function body', () => {
      const validator = new RequiredNull({ optimize: false })
      expect(validator.codeGen('value1', 'validator1')).toMatchSnapshot()
    })

    it('toString, constructor', () => {
      const validator = new RequiredNull({ optimize: false })
      const code = validator.toString()
      expect(code).toEqual('new RequiredNull({ optimize: false })')
    })

    it('toString, typescript', () => {
      const validator = new RequiredNull({ optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual('null')
    })
  })
})

describe.each([false, true])('Null (optimize: %s)', optimize => {
  describe('RequiredNull', () => {
    it('should generate validation code and give same result', () => {
      const validator = new RequiredNull({ optimize })
      if (optimize) {
        expect(validator['optimizedValidate']).not.toBeNull()
      } else {
        expect(validator['optimizedValidate']).toBeNull()
      }
      const errors = validator.validate(null)
      expect(errors).toEqual([])
    })

    it('should export validator code with options', () => {
      const validator = new RequiredNull({ optimize })
      const code = validator.toString()
      if (optimize) {
        expect(code).toEqual('new RequiredNull()')
      } else {
        expect(code).toEqual('new RequiredNull({ optimize: false })')
      }
    })

    it('accepts valid values', () => {
      const validator = new RequiredNull({ optimize })
      expect(validator.validate(null)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, null>).toEqual(true)
    })

    it('rejects invalid values', () => {
      const validator = new RequiredNull({ optimize })
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(false)).toStrictEqual([new NotNullFail('Must be null', false)])
      expect(validator.validate(1)).toStrictEqual([new NotNullFail('Must be null', 1)])
      expect(validator.validate(123.9)).toStrictEqual([new NotNullFail('Must be null', 123.9)])
      expect(validator.validate('1')).toStrictEqual([new NotNullFail('Must be null', '1')])
      expect(validator.validate('')).toStrictEqual([new NotNullFail('Must be null', '')])
      expect(validator.validate({})).toStrictEqual([new NotNullFail('Must be null', {})])
      expect(validator.validate([])).toStrictEqual([new NotNullFail('Must be null', [])])
      expect(true as AssertEqual<typeof validator.tsType, null>).toEqual(true)
    })

    it('rejects undefined', () => {
      const validator = new RequiredNull({ optimize })
      expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required', undefined)])
    })
  })

  describe('OptionalNull', () => {
    it('accepts empty value', () => {
      const validator = new OptionalNull({ optimize })
      expect(validator.validate(undefined)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, null | undefined>).toEqual(true)
    })

    it('toString, constructor', () => {
      const validator = new OptionalNull({ optimize: false })
      const code = validator.toString()
      expect(code).toEqual('new OptionalNull({ required: false, optimize: false })')
    })

    it('toString, typescript', () => {
      const validator = new OptionalNull({ optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual('null | undefined')
    })
  })
})

describe('Rust Types', () => {
  it('Error', () => {
    expect(() => {
      new RequiredNull().toString({ types: true, language: 'rust' })
    }).toThrow(`Null is not supported in Rust`)
  })
})
