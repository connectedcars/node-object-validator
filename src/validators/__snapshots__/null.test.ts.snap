// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Null RequiredNull should return a function body 1`] = `
Array [
  Object {
    "NotNullFail": [Function],
    "RequiredFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 !== undefined) {",
    "  if (value1 !== null) {",
    "    errors.push(new NotNullFail(\`Must be null\`, value1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
